### PR TITLE
pump/: Accelerat transaction status queries throught GetMvccByEncodeKey

### PR DESCRIPTION
**Pull request by july2993**
### What problem does this PR solve? 
Avoid some case that we wait 10 minutes after the transaction and call `GetTxnStatus`  which send the [CleanupRequest](https://github.com/pingcap/kvproto/blob/d288a78c7482b8abf254b97ec87fad3128f6c424/proto/kvrpcpb.proto#L242) that will get the commit ts of transaction or rollback the transaction.

### What is changed and how it works?
Query txn info by [GetMvccByEncodeKey](https://github.com/pingcap/kvproto/blob/d288a78c7482b8abf254b97ec87fad3128f6c424/proto/kvrpcpb.proto#L460).
we can know if the txn if committed or rollback as quick as possible.

If we can't get the status we will fallback to the case call GetTxnStatus after 10 minute of the transaction.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
     * using sysbench tps = 10k
     * kill -9 tidb
  log of the pump with this pr
```
[tidb@ip-172-16-4-74 tikv]$ tail  -f log/pump.log | grep "known txn"
[2019/06/10 17:24:59.192 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244264] ["commit ts"=408986240913244423]
[2019/06/10 17:25:02.192 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244269] ["commit ts"=408986240913244494]
[2019/06/10 17:25:05.192 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244275] ["commit ts"=408986240913244488]
[2019/06/10 17:25:08.193 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244292] ["commit ts"=408986240913244482]
[2019/06/10 17:25:11.193 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244298] ["commit ts"=408986240913244518]
[2019/06/10 17:25:14.195 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244344] ["commit ts"=408986240913244468]
[2019/06/10 17:25:17.194 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244362] ["commit ts"=408986240913244485]
[2019/06/10 17:25:20.194 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244336] ["commit ts"=408986240913244517]
[2019/06/10 17:25:23.199 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244363] ["commit ts"=408986240913244419]
[2019/06/10 17:25:26.197 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244364] ["commit ts"=408986240913244447]
[2019/06/10 17:25:29.197 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244342] ["commit ts"=408986240913244489]
[2019/06/10 17:25:32.197 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244314] ["commit ts"=408986240913244448]
[2019/06/10 17:25:35.201 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244379] ["commit ts"=408986240913244450]
[2019/06/10 17:25:38.200 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244391] ["commit ts"=408986240913244506]
[2019/06/10 17:25:41.200 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244384] ["commit ts"=408986240913244528]
[2019/06/10 17:25:44.200 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244397] ["commit ts"=408986240913244505]
[2019/06/10 17:26:16.494 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244401] ["commit ts"=0]
[2019/06/10 17:26:17.497 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244406] ["commit ts"=0]
[2019/06/10 17:30:21.015 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244404] ["commit ts"=0]
[2019/06/10 17:30:22.017 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244410] ["commit ts"=408986240913244542]
[2019/06/10 17:30:23.023 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244412] ["commit ts"=0]
[2019/06/10 17:30:24.026 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244418] ["commit ts"=408986240913244531]
[2019/06/10 17:30:25.028 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244428] ["commit ts"=0]
[2019/06/10 17:34:55.403 +08:00] [INFO] [storage.go:468] ["known txn is committed from tikv"] ["start ts"=408986240913244441] ["commit ts"=0]
[2019/06/10 17:34:56.406 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244455] ["commit ts"=0]
[2019/06/10 17:34:57.408 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244463] ["commit ts"=0]
[2019/06/10 17:34:58.411 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244460] ["commit ts"=0]
[2019/06/10 17:34:59.413 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244454] ["commit ts"=0]
[2019/06/10 17:35:00.415 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244497] ["commit ts"=0]
[2019/06/10 17:35:01.417 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244457] ["commit ts"=0]
[2019/06/10 17:35:02.423 +08:00] [INFO] [storage.go:468] ["known txn is committed from tikv"] ["start ts"=408986240913244474] ["commit ts"=0]
[2019/06/10 17:35:03.425 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244495] ["commit ts"=0]
[2019/06/10 17:35:04.437 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244479] ["commit ts"=0]
[2019/06/10 17:35:05.440 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244490] ["commit ts"=0]
[2019/06/10 17:35:06.442 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244453] ["commit ts"=0]
[2019/06/10 17:35:07.446 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244473] ["commit ts"=0]
[2019/06/10 17:35:08.448 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244512] ["commit ts"=0]
[2019/06/10 17:35:09.451 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244521] ["commit ts"=0]
[2019/06/10 17:35:10.531 +08:00] [INFO] [storage.go:433] ["known txn is committed or rollback from tikv"] ["start ts"=408986240913244523] ["commit ts"=0]
```

.74 is the version with this pr
Screen Shot 2019-06-10 at 5 56 59 PM

Screen Shot 2019-06-10 at 6 00 04 PM

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch
- Need to be included in the release note

### CODE REVIEW

File: pump/server.go [hunk 1]
```diff
@@ -597,6 +600,23 @@ func (s *Server) BinlogByTS(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprint(w, "\n\n PrewriteValue: \n")
 		fmt.Fprint(w, prewriteValue.String())
 	}
+
+	if len(binlog.PrewriteKey) > 0 {
+		tikvStorage := s.tiStore.(tikv.Storage)
+		healper := storage.Helper{
```
Comment (GregoryIan) [inline pump/server.go]: why not one `Helper` instance?

File: pump/server.go [hunk 2]
```diff
@@ -597,6 +600,23 @@ func (s *Server) BinlogByTS(w http.ResponseWriter, r *http.Request) {
...
+			Store:       tikvStorage,
+			RegionCache: tikvStorage.GetRegionCache(),
+		}
+
+		resp, err := healper.GetMvccByEncodedKey(binlog.PrewriteKey)
```
Comment (GregoryIan) [inline pump/server.go]: we can extract one function `queryTransactionStatus`, then just call it. the implementation output too many unless and  old mvcc records

**Comment by GregoryIan** [comment]
Good Job!

File: pump/storage/storage.go [hunk 3]
```diff
@@ -344,22 +358,92 @@ func (a *Append) updateStatus() {
...
+	} else {
+		for _, w := range resp.Info.Writes {
+			if !(int64(w.StartTs) == startTS) {
+				continue
+			}
+
+			if w.Type != kvrpcpb.Op_Rollback {
+				// Sanity checks
+				if int64(w.CommitTs) <= startTS {
+					log.Warn("op type not Rollback, but have unexpect commit ts",
```
Comment (GregoryIan) [inline pump/storage/storage.go]: there must be some critical problems with TiDB, you can log error log and return true? don't block replication progress

File: pump/storage/storage.go [hunk 4]
```diff
@@ -344,22 +358,92 @@ func (a *Append) updateStatus() {
...
+				}
+
+				err := a.writeCBinlog(pbinlog, int64(w.CommitTs))
+				if err != nil {
+					log.Error("writeCBinlog failed", zap.Error(err))
+					return false
+				}
+			} else {
+				// Will get the same value as start ts if it's rollback, set to 0 for log
+				w.CommitTs = 0
```
Comment (GregoryIan) [inline pump/storage/storage.go]: no need to change it

**Comment by GregoryIan** [comment]
```
if item.tp == pb.BinlogType_Prewrite {
	getTime := time.Now()

	for {
		_, ok := s.waitStartTS[item.start]
		if !ok {
			break
		}

		// we may get the C binlog soon at start up time
		if time.Since(getTime) > time.Second {
		        if s.resolver != nil && s.resolver(item.start) {
				break
			}
		}
```
should we put  `getTime := time.Now()` in the `for` loop?

File: pump/storage/storage.go [hunk 5]
```diff
@@ -344,22 +358,92 @@ func (a *Append) updateStatus() {
...
+	} else if len(resp.Error) > 0 {
+		log.Error("GetMvccByEncodedKey failed", zap.String("Error", resp.Error))
+	} else {
+		for _, w := range resp.Info.Writes {
+			if !(int64(w.StartTs) == startTS) {
+				continue
+			}
+
+			if w.Type != kvrpcpb.Op_Rollback {
+				// Sanity checks
```
Comment (GregoryIan) [inline pump/storage/storage.go]: does `w.CommitTs == 0`  imply `still not commited`? should we check it firstly?

File: pump/storage/storage.go (hunk 4 above)
Comment (july2993) [reply pump/storage/storage.go]: easy to see whether it is commited (o rollback), and keep it same as the log of L471( the case call `GetTxnStatus`)

File: pump/storage/storage.go (hunk 3 above)
Comment (july2993) [reply pump/storage/storage.go]: chang to error and break to fallback to the old logit calling GetTxnStatus if needed for safety.

File: pump/server.go (hunk 1 above)
Comment (july2993) [reply pump/server.go]: no performance cost and only use here so I think better just construct here, don't make Server struct complecate

File: pump/server.go (hunk 2 above)
Comment (july2993) [reply pump/server.go]: for debug just keep all info?

File: pump/storage/helper.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// Helper is a middleware to get some information from tikv/pd.
+type Helper struct {
+	Store       tikv.Storage
+	RegionCache *tikv.RegionCache
+}
+
+// GetMvccByEncodedKey get the MVCC value by the specific encoded key.
+func (h *Helper) GetMvccByEncodedKey(encodedKey kv.Key) (*kvrpcpb.MvccGetByKeyResponse, error) {
```
[the remark this answers is not in the document]
Comment (july2993) [reply pump/storage/helper.go]: // Copy from https://github.com/pingcap/tidb/blob/71def9c7263432c0dfa6a5960f6db824775177c9/store/helper/helper.go#L47
// we can use it directly if we upgrade to the latest version of TiDB dependency.

add comment in helper.go
and we may add other myself if not use TiDB code directly.

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**